bee-network/bee-autopeering/src/peering/neighbor.rs

```diff
@@ -0,0 +1,347 @@
+        }
+    }
+
+    fn is_preferred(&mut self, candidate: &Neighbor) -> bool {
+        if let Some(furthest) = self.find_furthest() {
+            candidate < furthest
+        } else {
+            true
+        }
+    }
+
+    fn select_from_candidate_list<'a>(&mut self, candidates: &'a [&'a Neighbor]) -> Option<&'a Peer> {
+        if candidates.is_empty() {
+            None
+        } else if let Some(furthest) = self.find_furthest() {
+            for candidate in candidates {
+                if *candidate < furthest {
+                    return Some(candidate.peer());
+                }
+            }
+            None
+        } else {
+            // Any candidate can be selected: pick the first.
+            Some(candidates[0].peer())
+        }
+    }
+
+    fn find_furthest(&mut self) -> Option<&Neighbor> {
+        if self.neighbors.len() >= N {
+            self.neighbors.sort_unstable();
+            self.neighbors.last()
+        } else {
+            None
+        }
+    }
+
+    fn remove_furthest(&mut self) -> Option<Peer> {
+        // Note: Both methods require unique access to `self`, so we need to copy the peer id.
+        if let Some(peer_id) = self.find_furthest().map(|d| *d.peer().peer_id()) {
+            self.remove_neighbor(&peer_id)
+        } else {
+            None
+        }
+    }
+
+    fn update_distances(&mut self, local: &Local) {
+        let local_id = local.peer_id();
+        let salt = if INBOUND {
+            local.private_salt().expect("missing private salt")
+        } else {
+            local.public_salt().expect("missing public salt")
+        };
+
+        self.neighbors.iter_mut().for_each(|pd| {
+            pd.distance = salt_distance(&local_id, pd.peer().peer_id(), &salt);
+        });
+    }
+
+    fn len(&self) -> usize {
+        self.neighbors.len()
+    }
+
+    fn is_full(&self) -> bool {
+        self.neighbors.len() == N
+    }
+
+    fn clear(&mut self) {
+        self.neighbors.clear();
+    }
+}
+
+impl<const N: usize, const INBOUND: bool> Default for NeighborhoodInner<N, INBOUND> {
+    fn default() -> Self {
+        Self {
+            neighbors: Vec::with_capacity(N),
+        }
+    }
+}
+
+impl<const N: usize, const INBOUND: bool> fmt::Display for NeighborhoodInner<N, INBOUND> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}/{}", self.neighbors.len(), N)
+    }
+}
+
+// hive.go:
+// returns the distance (uint32) between x and y by xoring the hash of x and (y + salt):
+// xor( hash(x), hash(y+salt) )[:4] as little-endian uint32
+pub(crate) fn salt_distance(peer1: &PeerId, peer2: &PeerId, salt: &Salt) -> Distance {
+    let hash1 = hash::sha256(peer1.id_bytes());
+    let hash2 = hash::sha256(&concat(peer2.id_bytes(), salt.bytes()));
+
+    let xored = xor(hash1, hash2);
+
+    Bytes::copy_from_slice(&xored[..4]).get_u32_le()
+}
+
+fn concat<const N: usize, const M: usize>(bytes1: &[u8; N], bytes2: &[u8; M]) -> Vec<u8> {
```

Comment:
Can't this return an array of size M + N to avoid the Vec allocation ?

Reply:
IIRC unfortunately not. It's not that I didn't try. I was also confused why Rust wouldn't allow me to write `const l: usize = N + M;` (that Christian asked) which would make perfect sense, right?

I can't quite remember why, but for some reason you can only evaluate the symbol `N + M` once, and since you cannot store it in a const, you cannot create an array, which requires a const length parameter. It's definitely that at some point should be possible, but const generics doesn't seem to be there yet.

Reply:
there is something else we could do:
```rust
fn concat<const N: usize, const M: usize, const L: usize>(bytes1: &[u8; N], bytes2: &[u8; M]) -> &[u8; L] {
    const _: () = assert!(N + M == L);
    ...
```

Reply:
I'ld rather accept the extra allocation and keep the code simple, but let's make it a CAT's call I would say.

Autopeering is not something we need to optimize to death tbh. It's a relatively slow process doing something on a second timescale. If possible I'ld rather keep the code simple and easy to understand. @thibault-martinez Wdyt?

Reply:
Closing this due to lack of response :trollface: